R/survey_weights.R

```diff
@@ -0,0 +1,227 @@
+
+
+add_weights <- function(x, p, cz = NULL, ... , population,
+                        cluster_cz, household_cz,
+                        cluster_x, household_x, individuals_eligible_x, individuals_interviewed_x,
+                        method = "stratified",
+                        ignore_cluster = TRUE, ignore_household = TRUE,
+                        surv_weight = "surv_weight", surv_weight_ID =  "surv_weight_ID") {
+
+  # stratified method is straight forward - lifted directly from old function version
+  if (method == "stratified") {
+    .dots      <- rlang::enquos(...)
+    population <- rlang::enquo(population)
+    surv_weight_ID  <- rlang::enquo(surv_weight_ID)
+    surv_weight     <- rlang::enquo(surv_weight)
+
+    # create a merger ID by age group and sex
+    p <- tidyr::unite(p, "ID", !!! .dots)
+
+    # get study sample counts
+    counts <- dplyr::count(x, !!! .dots)
+    counts <- tidyr::unite(counts, "ID", !!! .dots)
+    counts <- dplyr::select(counts, .data$ID, .data$n)
+
+    # merge counts to population data
+    p <- dplyr::left_join(p, counts, by = "ID")
+
+    # create weight variable
+    p <- dplyr::mutate(p, !! surv_weight := !! population / .data$n)
+    p <- dplyr::select(p, .data$ID, !! surv_weight)
+
+
+    # merge to study sample
+    merge_by <- "ID"
+    names(merge_by) <- as.character(rlang::expr(surv_weight_ID))
+    d <- tidyr::unite(x, !! surv_weight_ID, !!! .dots, remove = FALSE)
+    d <- dplyr::left_join(d, p, by = merge_by)
+    d
+  }  else {
+
+    # cluster method
+
+    # define vars
+    clus_id_cz <- tidyselect::vars_select(colnames(cz), {{cluster_cz}})
+    hh_id_cz   <- tidyselect::vars_select(colnames(cz), {{household_cz}})
+    clus_id_x <- tidyselect::vars_select(colnames(x), {{cluster_x}})
+    hh_id_x   <- tidyselect::vars_select(colnames(x), {{household_x}})
+    indiv_eli   <- tidyselect::vars_select(colnames(x), {{individuals_eligible_x}})
+    indiv_surv   <- tidyselect::vars_select(colnames(x), {{individuals_interviewed_x}})
+
+
+    if (ignore_cluster) {
+      cluster_chance <- 1
+    } else {
+      ## chance of a cluster being selected
+      # the total number of clusters in the area
+      n_clusters_total  <- summarise(cz, nclus_tot = sum(!duplicated(clus_id_cz)))
+      # the number of clusters that were surveyed
+      n_clusters_chosen <- summarise(x, nclus = sum(!duplicated(clus_id_x)))
+      # inverse chance of a cluster being chosen
+      cluster_chance <- as.numeric(n_clusters_total / n_clusters_chosen)
+    }
+
+    if (ignore_household) {
+      c$hh_chance <- 1
+    } else {
+      ## chance of a household being selected in each cluster
+      # number of households chosen in each cluster
+      n_households_chosen <- summarise(
+        group_by(x, cluster_id_x),
+        n_hh = sum(!duplicated(hh_id_x))
+        )
+
+      # join counts of households surveyed to list of clusters
+      cz <- left_join(cz, n_households_chosen, by = c(clus_id_cz = clus_id_x))
```

Comment:
This will make sure you have the correct named vector for joining

```suggestion
      cz <- left_join(cz, n_households_chosen, by = setNames(clus_id_x, clus_id_cz))
```